src/lib.rs

```diff
@@ -187,20 +194,36 @@ impl<D> EventSource for WaylandSource<D> {
             }
         }
 
-        // ensure we are not waiting for messages that are already in the buffer.
-        Self::loop_callback_pending(&mut self.queue, &mut callback)?;
-        self.read_guard = Some(Self::prepare_read(&mut self.queue)?);
-
-        Ok(())
+        self.read_guard = self.queue.prepare_read();
+        match self.read_guard {
+            Some(_) => Ok(None),
+            // If getting the guard failed, that means that there are
+            // events in the queue, and
+            // The readiness value is never used, we just need some marker
+            // If getting the guard failed, we need to process the events 'instantly'
+            // tell calloop this
+            None => Ok(Some((Readiness::EMPTY, self.fake_token.unwrap()))),
+        }
     }
 
-    fn post_run<F>(&mut self, _: F) -> calloop::Result<()>
-    where
-        F: FnMut((), &mut Self::Metadata) -> Self::Ret,
-    {
-        // Drop implementation of ReadEventsGuard will do cleanup
-        self.read_guard.take();
-        Ok(())
+    fn before_handle_events(&mut self, events: calloop::EventIterator<'_>) {
+        let guard = self.read_guard.take();
+        if events.count() > 0 {
+            // 1. read events from the socket if any are available
+            if let Some(guard) = guard {
+                if let Err(WaylandError::Io(err)) = guard.read() {
+                    // If some other thread read events before us, concurrently, that's an expected
+                    // case, so this error isn't an issue. Other error kinds do need to be returned, however
+                    if err.kind() != io::ErrorKind::WouldBlock {
+                        // before_handle_events doesn't allow returning errors
+                        // For now, cache it in self until process_events is called
+                        self.stored_error = Some(err);
+                    }
+                }
+            }
+        }
+        // Otherwise, drop the guard if we have it, as we don't want to do any
+        // reading when we didn't get any events
```

Comment:
it's always dropped? There's no otherwise.

Reply:
Well it's always dropped, sure. But it's not always dropped by us - if the `read` call happens, that takes ownership and drops it.

Reply:
If we call the `read` it still sort of dropped by us. The comment is a bit misleading imho, because `read` should take ownership, no?